File: src/面试题08_09.bracket-lcci.100200/solution.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    const result = [];
    backTracking(n, n, [], result);
    return result;
};

function backTracking (leftCount, rightCount, sequence, result) {
    if (leftCount === 0 && rightCount === 0) {
        result.push(sequence.join(''));
        return;
    }
    if (leftCount > 0) {
        sequence.push('(');
        backTracking(leftCount - 1, rightCount, sequence, result);
        sequence.pop();
    }

    if (rightCount > leftCount) {
        sequence.push(')');
        backTracking(leftCount, rightCount - 1, sequence, result);
        sequence.pop();
    }
}

File: src/面试题08_09.bracket-lcci.100200/solution.java
class Solution {
    private final List<String> result = new ArrayList<>();
    private final StringBuilder sb = new StringBuilder();
    public List<String> generateParenthesis(int n) {
        backTracking(n,n);
        return result;
    }

    private void backTracking(int left, int right){
        if(left == 0 && right == 0){
            result.add(sb.toString());
            return;
        }
        
        if(left>0){
            sb.append('(');
            backTracking(left-1,right);
            sb.deleteCharAt(sb.length()-1);
        }
        
        if(right>left){
            sb.append(')');
            backTracking(left,right-1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}